state/traverse.go

```diff
@@ -7,27 +7,60 @@ import (
 	"github.com/onflow/flow-go/storage"
 )
 
-// Traverse traverses a chain segment beginning with the start block (inclusive)
-// and ending with the end block (inclusive). Blocks are traversed in reverse
+// functor that will be called on each block header when traversing blocks.
+type onVisitBlock = func(header *flow.Header) error
+
+// functor that will be called on each block header to know if we should continue traversing the chain.
+type shouldContinue = func(header *flow.Header) bool
+
+// TraverseBackward traverses a chain segment beginning with the start block (inclusive)
+// Blocks are traversed in reverse
 // height order, meaning the end block must be an ancestor of the start block.
 // The callback is called for each block in this segment.
-func Traverse(headers storage.Headers, start, end flow.Identifier, callback func(header *flow.Header) error) error {
-
-	nextID := start
+// Return value of callback is used to decide if it should continue or not.
+func TraverseBackward(headers storage.Headers, startBlockID flow.Identifier, visitor onVisitBlock, shouldContinue shouldContinue) error {
+	blockID := startBlockID
+	// in case we reached genesis
 	for {
-		// retrieve the next block in the segment and pass it to the callback
-		next, err := headers.ByBlockID(nextID)
+		block, err := headers.ByBlockID(blockID)
 		if err != nil {
-			return fmt.Errorf("could not get segment block (id=%x): %w", nextID, err)
+			return fmt.Errorf("could not get block header (%x): %w", blockID, err)
 		}
-		err = callback(next)
+
+		err = visitor(block)
 		if err != nil {
-			return fmt.Errorf("error in callback: %w", err)
+			return err
 		}
 
-		if nextID == end {
+		if !shouldContinue(block) {
 			return nil
 		}
-		nextID = next.ParentID
+
+		blockID = block.ParentID
+	}
+}
+
+// TraverseForward traverses a chain segment in forward order.
+// Implements a recursive descend starting at the `forkHead` towards the genesis block. The descend continues as long as `shouldContinue` returns true. All visited blocks, which `shouldContinue` returned true for, are fed into the `visitor` callback starting from the block with the lowest height, in order of increasing height. The last block that is fed into `visitor` is the `forkHead`.
+func TraverseForward(headers storage.Headers,
+	forkHead flow.Identifier,
+	visitor onVisitBlock,
+	shouldContinue shouldContinue,
+) error {
+	var blocks []*flow.Header
+	err := TraverseBackward(headers, forkHead, func(header *flow.Header) error {
+		blocks = append(blocks, header)
+		return nil
+	}, shouldContinue)
+	if err != nil {
+		return err
+	}
+
+	for i := range blocks {
+		err = visitor(blocks[len(blocks)-i-1])
```

Comment:
to me, this loop would be more intuitive when written as:


```suggestion
         i := len(blocks) -1
	for i >= 0 {
		err = visitor(blocks[i])
		i--
```